### PR TITLE
Bratseth/upgrade to latest good

### PR DESCRIPTION
**Pull request by bratseth**
@mpolden please review

### CODE REVIEW

**Comment by mpolden** [comment]
Looks good, but can't this cause a downgrade, which is unsupported?

Consider:
v2 = low
v1 = normal

v2 = normal - canaries complete
v1 = normal
An application begins upgrading to v2, a subset of nodes upgrade, but does not complete.

v2 = broken - failures results in broken version
v1 = normal
The application downgrades to v1.

If we change `notFailingOn` to be `notFailingOnEqualOrHigher` it might be fine.

**Comment by bratseth** [comment]
Good point. It can't happen in one zone, because we only schedule an upgrade if a deployment is on an older version according to the deployed application, regardless of whether that production job failed, but it can happen if one zone gets to this point while another is on an older version. It is an exotic state but we should protect against it. I've added double protection in the last commit:
- Don't trigger a production upgrade job if the last completed is on a newer version.
- Fail deployments that downgrade.

We still want to schedule upgrades even if some production deployment zone is on a newer version if another is on an older version, to get rid of that older version.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 1]
```diff
@@ -404,6 +416,40 @@ private boolean isRunningProductionJob(Application application) {
         return application.deploymentJobs().jobStatus().entrySet().stream()
                 .anyMatch(entry -> entry.getKey().isProduction() && entry.getValue().isRunning(jobTimeoutLimit()));
     }
+
+    /**
+     * When upgrading it is ok to trigger the next job even if the previous failed if the previous has earlier succeeded
+     * on the version we are currently upgrading to
+     */
+    private boolean isProductionJobPreviouslySucceedingOnCurrentlyDeployingVersion(JobStatus jobStatus, Change change) {
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: Consider something like `productionJobHasSucceededFor(change, jobStatus)`

and javadoc:
"Returns true if a production job has already succeeded for the given change"

Naming things is difficult :-)

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java [hunk 2]
```diff
@@ -404,6 +416,40 @@ private boolean isRunningProductionJob(Application application) {
...
+        if ( ! (change instanceof Change.VersionChange) ) return false;
+        if ( ! isProduction(jobStatus.type())) return false;
+        Optional<JobStatus.JobRun> lastSuccess = jobStatus.lastSuccess();
+        if ( ! lastSuccess.isPresent()) return false;
+        return lastSuccess.get().version().equals(((Change.VersionChange)change).version());
+    }
+
+    /** 
+     * Returns whether the current deployed version in the zone given by the job
+     * is never than the given version. This may be the case even if the production job
```
Comment (mpolden) [inline controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: "never" -> "newer"

**Comment by mpolden** [comment]
Also, controller-server failed to compile.

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java (hunk 1 above)
Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: That works, thanks!

File: controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java (hunk 2 above)
Comment (bratseth) [reply controller-server/src/main/java/com/yahoo/vespa/hosted/controller/deployment/DeploymentTrigger.java]: fixed

**Comment by bratseth** [comment]
The PR build error is spurious, merging.

**Pull request merged**